v2/api/devices/customizations/iot_hub_extension.go

```diff
@@ -0,0 +1,162 @@
+	// the hub type has been changed but this extension has not
+	var _ conversion.Hub = typedObj
+
+	hasSecrets := secretsSpecified(typedObj)
+	if !hasSecrets {
+		log.V(Debug).Info("No secrets retrieval to perform as operatorSpec is empty")
+		return nil, nil
+	}
+
+	id, err := genruntime.GetAndParseResourceID(typedObj)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make(map[string]armiothub.SharedAccessSignatureAuthorizationRule)
+	// Only bother calling ListKeys if there are secrets to retrieve
+	if hasSecrets {
+		subscription := id.SubscriptionID
+		// Using armClient.ClientOptions() here ensures we share the same HTTP connection, so this is not opening a new
+		// connection each time through
+		var resClient *armiothub.ResourceClient
+		resClient, err = armiothub.NewResourceClient(subscription, armClient.Creds(), armClient.ClientOptions())
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create new DevicesClient")
+		}
+
+		var pager *runtime.Pager[armiothub.ResourceClientListKeysResponse]
+		var resp armiothub.ResourceClientListKeysResponse
+		pager = resClient.NewListKeysPager(id.ResourceGroupName, typedObj.AzureName(), nil)
+		for pager.More() {
+			resp, err = pager.NextPage(ctx)
+			addSecretsToMap(resp.Value, keys)
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to retreive response")
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed listing keys")
+		}
+
+	}
+
+	secretSlice, err := secretsToWrite(typedObj, keys)
+	if err != nil {
+		return nil, err
+	}
+
+	return secrets.SliceToClientObjectSlice(secretSlice), nil
+}
+
+func secretsSpecified(obj *storage.IotHub) bool {
+	if obj.Spec.OperatorSpec == nil || obj.Spec.OperatorSpec.Secrets == nil {
+		return false
+	}
+
+	secrets := obj.Spec.OperatorSpec.Secrets
+
+	if secrets.IotHubOwnerPrimaryKey != nil ||
+		secrets.IotHubOwnerSecondaryKey != nil ||
+		secrets.ServicePrimaryKey != nil ||
+		secrets.ServiceSecondaryKey != nil ||
+		secrets.RegistryReadWritePrimaryKey != nil ||
+		secrets.RegistryReadWriteSecondaryKey != nil ||
+		secrets.RegistryReadPrimaryKey != nil ||
+		secrets.RegistryReadSecondaryKey != nil ||
+		secrets.DevicePrimaryKey != nil ||
+		secrets.DeviceSecondaryKey != nil {
+		return true
+	}
+
+	return false
+}
+
+func addSecretsToMap(keys []*armiothub.SharedAccessSignatureAuthorizationRule, result map[string]armiothub.SharedAccessSignatureAuthorizationRule) {
+	for _, key := range keys {
+		if key == nil || key.KeyName == nil {
+			continue
+		}
+		result[*key.KeyName] = *key
+	}
+}
+
+func secretsToWrite(obj *storage.IotHub, keys map[string]armiothub.SharedAccessSignatureAuthorizationRule) ([]*v1.Secret, error) {
+	operatorSpecSecrets := obj.Spec.OperatorSpec.Secrets
+	if operatorSpecSecrets == nil {
+		return nil, errors.Errorf("unexpected nil operatorspec")
+	}
+
+	collector := secrets.NewCollector(obj.Namespace)
+	iothubowner, ok := keys["iothubowner"]
```

Comment:
Where did you get these values? Is it in some IOTHub documentation? Can you link it here? Or did you just get it by observing the payload of the response?

Reply:
Link : https://learn.microsoft.com/en-us/rest/api/iothub/iot-hub-resource/list-keys?tabs=HTTP#sharedaccesssignatureauthorizationrule

Dropping a comment in code as well